### ISSUE TITLE
[RFC]: Dynamically loaded components

### ISSUE THREAD

**Issue by mykas** [wix/wix-style-react#4847]
## [RFC]: Dynamically loaded components

* [The Motivation](#the-motivation)
* [The Proposal](#the-proposal)
   * [Candidate](#first-candidate)
   * [Technology](#technology)
* [Issues to Tackle](#issues-to-tackle)
   * [Smart preloading](#smart-preloading)   
   * [Testing patterns](#testing-patterns)
   * [ES modules](#es-modules)
* [Expected Wins](#expected-wins)
* [Next Steps](#next-steps)

## <a href="#the-motivation" id="the-motivation">#</a> The Motivation

Wix would like to improve page loads on backoffice applications. By making perfer mandatory - now almost every team including us can measure bundle size and site metrics. This opens up posibility to improve gradually our bundle sizes in measured way. 

#### Industry Standards

The target page load time in the industry is expected to be below 3s. 

Every second matters to our user experience . Here is an example from google where they measured the bounce probability based on how many seconds it took to load the app on mobile.  
![image](https://user-images.githubusercontent.com/8919190/72185506-9d794600-33fb-11ea-9281-7aa508ac919a.jpg)

In short, speed equals revenue.

>Keep in mind that the time is measured in when application network is idle for 500ms, which means that it includes not just js bundles but images too. 

#### Where can we have impact

One of the most impactful parts of the app loading time is javascript bundles. The bigger the size - the longer it takes to download it. This is where we can improve and have impact on all verticals.

#### Backoffice metrics

Almost all backoffice applications are developed with our components which means we have huge impact on their bundles size.

Here are some examples:

 
`Events.Dashboard`. Main dashboard bundle: 604 KB (Gzip). Wix-style-react:  118 KB(Gzip) + all required dependencies like popper.js, draft-js etc. So it means even more than 118KB.

`Promote.Shoutout-Analytics`. Newest project. App bundle: 301 KB (Gzip). Wix-style-react:  66 KB(Gzip) + all required dependencies like popper.js etc. So it means even more than 66KB.

`Members. Backoffice`. Main bundle: 494 KB (Gzip). Wix-style-react: 90 KB + all required dependencies like popper.js, draft-js etc. So it means even mnore then 90 KB.

> All examples are calculated with NODE_EVN=production enabled on yoshi which means bundles are minified.

## <a href="#the-proposal" id="the-proposal">#</a> The Proposal

I believe it's possible to reduce initial bundle size by introducing dynamic loading to some of our heaviest components like: Popover, RichtTextInputArea and others that include quite heavy parts.By reducing initial bundle we decrease the loading time for our business manager apps. 

Let us take an example of Events page. As we already mentioned their main bundle is 604 KB (Gzip). This bundle is being preloaded when Events button is clicked on the sidebar.  Until bundle gets loaded + parsed it will show a loader.
![Events_loading](https://user-images.githubusercontent.com/8919190/72249970-3928d300-3603-11ea-8c2c-6dbadfa7666d.gif)

I know for certain that their main page uses two biggest components of our library: Popover & RichtTextInputArea. It is quite a heavy bundle and making impact on such bundle is pretty hard, but let us assume that event team decides to cut the initial page bundle size to 200 Kb (Gzip) and introduce dynamic loading for different tabs. 

How can we impact with dynamic loading:

* If we introduce 15Kb reduction on our Popover component - it will be 7.5 % load decrease for their initial page.
* If we introduce a 100 Kb reduction for our RichTextInputArea - it will be 50 % load decrease for their initial page. 

#### <a href="#first-candidate" id="first-candidate">#</a> First candidate

Our Popover is one of the heaviest (24.5Kb) and most used component in our library. It would make sense to have dynamic loading in this component and to see the big change all across the library. From investigation I managed to figure out that it is possible to cut by half the size just by dynamically loading the popover content element which bundles: `popper.js`, `react-transition-group`, `react-portal` libraries.

#### <a href="#technology" id="technology">#</a> Technology

Artem and later on I did some research on whats available in the market for dynamic loading and most of the solutions are `dynamic imports + promises approach`. In order not to break our users - we need a solution that supports both: async and sync modes. 

This can be achieved already with Artem Yavorsky solution called Loadable (wix-ui-core). This solution enables us to control the way we want build our component (with dynamic loading or without). The feature of dynamic loading would only work in production in order to avoid issues that could happen in `test` or `development` environments.

```jsx
    <LoadablePopper
      loader={{
        Popper:
          process.env.NODE_ENV === 'test' ||
          process.env.NODE_ENV === 'development'
            ? () => require('react-popper')
            : () => import('react-popper'),
        CSSTransition:
          process.env.NODE_ENV === 'test' ||
          process.env.NODE_ENV === 'development'
            ? () => require('react-transition-group')
            : () => import('react-transition-group'),
        Portal:
          process.env.NODE_ENV === 'test' ||
          process.env.NODE_ENV === 'development'
            ? () => require('react-portal')
            : () => import('react-portal'),
      }}
      namedExports={{
        Popper: 'Popper',
        CSSTransition: 'CSSTransition',
        Portal: 'Portal',
      }}
    >
```

## <a href="#issues-to-tackle" id="issues-to-tackle">#</a> Issues to tackle

Dynamic loading can work, but there are some technical issues we need to tackle.

#### <a href="#smart-preloading" id="smart-preloading">#</a> Smart preloading

I have already shown an example of how can we impact our users bundles by reducing the initial bundle, but we need to think of other smart ways to load the popovers content only when it is needed and without delay. 

Some examples:
* Tooltip should have already preloaded content element chunk when user hovers on the trigger element otherwise it will create "lagging" feeling. 

* Popover based components like `Dropdowns` can load its dropbox (popover content) on user action like click or keyboard click. So in this case we can preload the content only when user interacts with it.

*  EllipsisHOC could preload its chunk only when its trigger element is in ellipsed mode(the three dots mode). 

All of these cases works well, but if only one of them already preloaded chunk - it will not make sense on other cases. What it means that if  all three or at least two of these cases are together used on the same page - it will not have additional benefit other than initial bundle size reduction. 

The bottle neck is the "lagging" feeling for components that require immediate response, but for fast speed internets this `"lagging" feeling is not present.  Our browsers these days are pretty smart regarding detecting the network speed - we could pretty much do the smart preloading only on fast speed networks.

#### <a href="#testing-patterns" id="testing-patterns">#</a> Testing patterns

We provide testkits to our users and one of the risks that we can have with dynamic loading is when our users will test our dynamically loaded component in production mode. 

Lets say we have a component like this:
```
<Popover
   shown={true}
>
 <Popover.Content>Content</Popover.Content>
 <Popover.Element><Butto>element</Button></Popover.Content>
</Popover>
```
As simple as `popoverTestkit.isContentElementExists()` would return false if called immediately and not awaited with `wix-eventually` module. Until now it worked without `eventually` but now our users will really need to wait until dynamically loaded chunk will be loaded.

#### <a href="#es-modules" id="es-modules">#</a> ES modules

Dynamic loading only works if we have fully functional es-modules infrastructure. 

For now we still recommend our users to import by picking components: `import Popover from 'wix-style-react/Popover'`, but those components are commonjs and `imports()` of dynamically loaded component would be transpiled to simple `require()`. 

So we need to make sure to migrate our users to use named imports `import { Popover } from 'wix-style-react'`. 

## <a href="#expected-wins" id="expected-wins">#</a> Expected Wins

1. Dynamically loaded Popover will reduce around `14Kb` from its original bundle size. Which then will affect all other components like Tooltip, EllipsisHOC, PopoverMenu and even Dropdowns etc.
![Screenshot 2020-01-09 at 16 11 37](https://user-images.githubusercontent.com/8919190/72075226-fcf02c80-32fb-11ea-91fc-61a90393e4b6.png)

2. As a side effect our library will have working `tree shaking` which will reduce even more Kb's from our users bundles.

3. We will have working solution as an example to do the same on other components.

### Phase 1

* [x] implement PopoverNext in wix-ui-core that will have dynamic loading working for testing purposes
* [x] implement TooltipNext in wix-ui-core that will use PopoverNext for testing purposes.
* [x] migrate withEllipsisHOC to use TooltipNext
* [x] migrate Popover in wix-style-react to use PopoverNext
* [x] migrate Tooltip in wix-style-react to use TooltipNext

### Phase 2

* [x] ensure that es-modules and named imports are working in wix-style-react
* [x] introduce named-importing to our wix-ui-icons-common
* [x] introduce fallback element to PopoverNext
* [x] introduce error handling when chunk fails to load on PopoverNext
* [x] test with real client projects
* [ ] test with ssr projects

### Phase 3 
* [x] write migration guide and codemode to migrate users to use named imports
* [ ] (wix-style-react 8) - merge Popover and Tooltip and withEllipsedTooltip
* [ ]  (wix-ui-core 3) - merge PopoverNext, TooltipNext and withEllipsedTooltipNext
* [ ] be for the consumers when/if bugs come

**Comment by shlomitc** [comment]
Hey Mykolas, thanks for the detailed RFC. Here are some of the points we raised during our meeting today:

### Research Data
1. I really miss the bottom line and the numbers. After we do all that work, what is the expected bundle reduction and does it worth the effort?

### SSR
1. How will SSR work? does `loadable` take care of it?

### Testing
1. I have some concerns regarding testing depending on `NODE_ENV`. This means that we never test production code in our unit tests. Also, this means our tests would be different between `jest-jsdom` and `puppeteer`.
2. Will that have any effect on the `unidriver` as well?
3. Will consumers need to change their code? for example, will the dropdown methods will need to change? also for the "puppeteer" drivers.
4. Worth checking what happens in `jest-jsdom` environment without the `NODE_ENV` and how dynamic import works in node.

### Errors, Network Issues, and Recovery:
This is a topic @giladsegal raised and I think is crucial.
1. (Fallback) What should appear for each component while loaded? In the `Tooltip`, it might be ok not to show anything...
2. (Error Handling) What happens if the chunk fails to load? In the `Tooltip` it might not be that critical, but what if a `Dropdown` content doesn't load?

### Types
1. Need to make sure that the `wsr-types` support also the esm entry, otherwise we will lose the benefits of the types.

### Rollout plan
1. I'm missing information on how are we going to test it? which projects are going to be beta-testers of it? We will probably need to test client projects and also SSR.
2. Need to mention what is the minimal version we now support (webpack 4)
3. What kind of codemods do we plan on providing to ease the migration

### Other performance gains
1. It's not directly related, but there are some dependencies we can dynamically load like `date-fns` which might be a big dependency as well.

CC @simast @argshook @yavorsky @giladsegal

**Comment by mykas** [comment]
Thanks for all the feedback! 

Here is clarification on some points. More will come later.

## Testing

 1. yoshi is transpiling `imports` to `require` on `test` environments too: 
![Screenshot 2020-01-22 at 10 53 08](https://user-images.githubusercontent.com/8919190/72879410-7aa53680-3d05-11ea-9211-b06de360f6bb.png). 

Whats interesting though is that `development` environment does support dynamic loading, so maybe we should too?

2. No direct effect on `unidriver`

3. The only thing what users will need to do is have eventually on specific method called: `isContentElementExists` only when running in environment: `production`. Other then that it will be the same. Maybe codemode can be developed to change this automatically for our users.

4. Current situation with dynamic loading in jsdom: https://github.com/jsdom/jsdom/issues/2651 which states that it is not currently supported.

**Comment by mykas** [comment]
## Research Data

1. I have already showed what kind of bundle size reduction we have on our perfer stats (around 14Kb), but the real benefit (including working tree shaking) we will only see after we migrate at least one project to use dynamic loading + named imports.

For now it is hard to predict, but if we put aside the tree shaking part then for example for `Events` team it will be 14Kb from its initial bundle. Which is not a lot knowing that they have a bundle of `600Kb`.

**Comment by mykas** [comment]
## Rollout Plan

1. I am thinking of taking dynamic imports package version from PR and try to test it out on `Events`, `Promote` and `Members` team as we have them here physically. Run unit tests and sled test to check whats going on.
3. For now I have in mind these two: named-imports (codemode) && isContentElementExists wrapping with eventually, but I think I will have more information after testing out on projects.

**Comment by mykas** [comment]
## Types

1. So I've talked with @mosheKerbel and it seems he is taking wsr-types library improvements on next sprint. Can I help you @mosheKerbel somehow you to migrate the types to accept named-importing?

**Comment by mykas** [comment]
## Errors, Network Issues, and Recovery

I will work on these heavily on my next sprint. All points make sense and I think it needs to be solved.

**Comment by shlomitc** [comment]
I don't want to be a party pooper, but now that we have the exact numbers, it makes me really wonder if "dynamic import" is the best solution we can suggest.

It will reduce 14kb out of bundle size which is good, however, it will also introduce another level of complexity to our components and failure point (errors), harder debugging, testing, and the reward of slightly lighter bundle size might not worth it...

Some more thoughts:

1. Let's say that we already had this mechanism - what more can be improved? How much more can we reduce from a typical application? I know that this requires a bit more research, but I think it is necessary.

2. Since we're moving towards esmodules and treeshaking anyway - how much does this reduce the common bundle size? will it make the imports from `react-popper` to be lighter?

3. If we remove some dependencies like `css-transition-group`, would that make a good impact already? how difficult will it be removing it?

**Comment by giladsegal** [comment]
One more thing I realized yesterday:
You said we don't use `React.Suspense` because it doesn't support `SSR` but I think that the components we try to dynamically load anyway aren't needed in `SSR`.
For example, we can use `Suspense` and render nothing in `SSR` context, in the client we can dynamically load tooltips, dropdowns and so on.
Using this approach, we also get fallback UI / error handling

**Comment by mykas** [comment]
Thanks both for replies! I will gather a bit more intormation and write it here

**Comment by mykas** [comment]
## More bundle size reduction numbers

Tree shaking impact:

Events. Dashboard events-business-manager.bundle.min.js  644KB —> 615KB = 29Kb (gzip) 
Promote.Shoutout-Analytics  app.bundle.min.js 302Kb —> 279Kb = 23Kb (gzip)
shoutout-analytics-component.bundle.min.js 291Kb —> 271Kb = 20Kb (gzip)  Members. Membershipt-tpa
bm-membership.bundle.min.js 426.691kb —> 406Kb (gzip) = 21 Kb (gzip)
connect-modal.bundle.min.js 279.32kb —> 262 Kb (gzip) = 17,32 Kb (gzip)
plan-setup-modal.bundle.min.js 272.811kb —> 254.46 Kb = 18.35 Kb (gzip)
membership-panel-widget.bundle.min.js 149.002kb  —> 121 kb = 28 Kb (gzip)

**Comment by shlomitc** [comment]
This is great!
We should definitely proceed in this direction.
Let's try to get the tree-shaking production-ready (incl. codemodes and types)

**Comment by mykas** [comment]
## Errors, Network Issues, and Recovery

So, after investigating the possibilities to implement custom error on failed chunk load and recovery mechanism here are the results:

**It is possible to show specific error and implement recovery mechanism**
![Popover_recovery](https://user-images.githubusercontent.com/8919190/73756895-fd4fdc00-4770-11ea-976d-0c04922bc841.gif)

**But there some issues that we need to think about:**

- if there are multiple popovers on the web page that are opened by default on page render - all of them will go to error state.
- recovery is possible through a button click, but user will require to click on all components one by one.
- layout of the page would look horrible if there are multiple popover that has failed state.
- during click to recover - we would need to show some kind of loading state. Question to ux probably: how should it look.

![Retrying_module](https://user-images.githubusercontent.com/8919190/73753613-a8f62d80-476b-11ea-82d1-e1ae66713709.gif)

Theoretically we can say that situations like opened popovers or dropdowns are quite rare so these kind of concerns are minimal, but this rises some concerns if all of this is actually worth the effort.

@shlomitc @giladsegal

**Comment by mykas** [comment]
So after thinking about my own comments above I just got this realisation that almost all backoffice applications do not handle their failed chunk cases like we try to do on our Popover case.

Here is an example of premium subscriptions page when one of their modals chunk has failed to load:

![Example_failure](https://user-images.githubusercontent.com/8919190/73823698-7e56b400-4801-11ea-9848-83805df6cbfd.gif)

As seen in gif - first I open with working chunk load and then I refresh the page when the chunk request is blocked by dev tools. The whole page just fails without any errors. Just blank.

So what I want to say that I believe that our Popover component could be first of a kind that would show our developers how should it be handled when used with dynamic loading.

**Comment by giladsegal** [comment]
The main difference is that verticals have the power to control what should
happen when an error occurs
They can display an error page that is related to their product and so on
A single component is out of context and this is why it is harder to
display fallback UI
I think that if we would use `React.Suspense` the client can provide it's
own error state handling out of the box

>

**Comment by mykas** [comment]
Hey @giladsegal , I actually did use now `React.Suspense`. It does not handle errors. We need to wrap it with Error Boundaries. React Suspense does not even handle recovery mechanism. Here is a good that I managed to find on how to handle it:
```
import * as React from 'react';

export function LazyLoaderFactory<T = any>(
  promise: () => Promise<{ default: React.ComponentType<any> }>,
  Loader = () => <>Loading...</>,
  ErrorRetryView = ({ retry }) => (
    <button onClick={retry}>Module loading error. Try again.</button>
  ),
) {
  function LazyLoader(props: T) {
    const [loading, setLoading] = React.useState<boolean>(true);
    const retry = React.useCallback(() => setLoading(true), []);
    const Lazy = React.useMemo(
      () =>
        React.lazy(() =>
          promise().catch(() => {
            setLoading(false);
            return { default: () => <ErrorRetryView retry={retry} /> };
          }),
        ),
      [promise, loading],
    );
    return (
      <React.Suspense fallback={<Loader />}>
        <Lazy {...props} />
      </React.Suspense>
    );
  }

  (LazyLoader as any).displayName = `LazyLoader`;

  return LazyLoader as React.ComponentType<T>;
}

```

**Comment by mykas** [comment]
Here is a message from Dan about recovery mechanism in Suspense: https://github.com/facebook/react/issues/14254#issuecomment-455929142

**Comment by giladsegal** [comment]
Actually the client can provide the error boundary just as regular suspense
code

>

**Comment by mykas** [comment]
What I mean about code recovery is that React.Suspense does not have `refetch` mechanism on some action. If ask our verticals to handle this - they would probably need to write something like: `refresh the page`. And even now they are not doing it - the page just crashes.

**Comment by giladsegal** [comment]
So instead of a blank page, they can provide a useful error page or use
some refetching mechanism (maybe we can provide something for them)

>

**Comment by mykas** [comment]
So what I am thinking about is basically to ask ourselves - can we trust our verticals to handle the errors or should we do it for them on our own components? 

Drawbacks as I said are that our errors could break the layout a bit, but its recoverable and it does not crash the whole page.

This idea of us handling the error has great benefits. For example if users is typing something in the form and in the middle of typing the whole page crashes - its super bad UX and usability wise. If it only crashes one component - user can still navigate through the page.

**Comment by mykas** [comment]
Also another issue with higher level boundaries that we could provide our users is that recovery from top level boundary is not possible. We can potentially remount the app on user action but triggering Reat.Lazy to refetch is problematic.

**Comment by shlomitc** [comment]
We can do some hybrid solution - provide a default error handling and allow custom error handling

**Comment by mykas** [comment]
yes, we can. I am glad that we are able to see these problems like `verticals not handling their failed chunk` because this is where we can come in and help them with proper backed solutions. I already talked with Domas about it and we will try to think about UX side of this on next sprint.

**Comment by mykas** [comment]
New update here: `React.Suspense` throws error during build time of SSR saying that ReactDom.Server does not support Suspense yet. I need to fallback using @loadable/component library for that.

**Comment by EgoziE** [comment]
Moved to Jira - closing
